### PR TITLE
SCALRCORE-17474 CSB > Set up migration tool (blobtruck) using Terraform

### PR DESCRIPTION
**Pull request by brezitskiy**

### CODE REVIEW

File: modules/google/variables.tf [hunk 1]
```diff
@@ -0,0 +1,43 @@
+variable "region" {
+  default = "us-central1"
+}
+
+variable "zone" {
+  default = "us-central1-a"
+}
+
+variable "machine_type" {
+  default = "f1-micro"
```
Comment (lbn) [inline modules/google/variables.tf]: Let's go with n1-standard-2 so that it's not shared.

File: modules/google/variables.tf [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+variable "database_instance_tier" {
+  default = "db-f1-micro"
```
Comment (lbn) [inline modules/google/variables.tf]: Could use a bigger instance here too by default.

File: modules/google/variables.tf [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  description = "The name of the bucket where sql dump of tf_blobs table is stored"
+}
+
+variable "sql_dump_file_name" {
+  type = string
+  description = "The name of sql dump file"
+}
+
+variable "max_allowed_packet" {
+  default = "131072"
```
Comment (lbn) [inline modules/google/variables.tf]: This is in bytes right? Needs to be 134217728.

File: modules/google/main.tf [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+    db_port = 3306
+    db_user = local.db_user
+    db_pass = local.db_pass
+    db_name = google_sql_database.database.name
+    crypto_key = var.crypto_key
+    bucket_id = var.target_bucket
+  }
+}
+
+resource "null_resource" "install_blobtruck_package" {
```
Comment (lbn) [inline modules/google/main.tf]: Could all this be done inside `google_compute_instance.node`?

File: modules/google/main.tf [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    host = google_compute_address.node_address.address
+    type = "ssh"
+    user = local.ssh_user
+    timeout = local.ssh_timeout
+    private_key = local.ssh_private_key
+  }
+
+  provisioner "remote-exec" {
+    inline = [
+      "gcloud sql import sql ${google_sql_database_instance.mysql.name} ${local.dump_uri} --database=default --quiet",
```
Comment (lbn) [inline modules/google/main.tf]: Could be done in the instance as well.

File: modules/google/main.tf [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+    content = data.template_file.blobtruck_yml.rendered
+    destination = "/var/tmp/blobtruck.yml"
+  }
+
+  provisioner "remote-exec" {
+    inline = [
+      "set -ex",
+      "sudo curl -s https://${var.packagecloud_token}:@packagecloud.io/install/repositories/scalr/scalr-server-ee-staging/script.rpm.sh | sudo bash",
+      "sudo yum --assumeyes install -q blobtruck",
+      "sudo cp /var/tmp/blobtruck.yml /etc/blobtruck/blobtruck.yml",
```
Comment (lbn) [inline modules/google/main.tf]: Maybe mv to avoid storing credentials in more than one place.

File: modules/google/main.tf [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+
+  provisioner "file" {
+    content = data.template_file.blobtruck_yml.rendered
+    destination = "/var/tmp/blobtruck.yml"
+  }
+
+  provisioner "remote-exec" {
+    inline = [
+      "set -ex",
+      "sudo curl -s https://${var.packagecloud_token}:@packagecloud.io/install/repositories/scalr/scalr-server-ee-staging/script.rpm.sh | sudo bash",
```
Comment (lbn) [inline modules/google/main.tf]: Need a `packagecloud_repo` variable which is set to `scalr-server-ee` by default.

File: modules/google/README.md [hunk 8]
```diff
@@ -0,0 +1,20 @@
+# Terraform GCP blobtruck module
+
+Configures blobtrunk for migrating tf-blob service data from local backend into the Google Cloud Storage
```
Comment (lbn) [inline modules/google/README.md]: blobtruck.

File: modules/google/README.md [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+| region | n/a | `string` | `"us-central1"` | no |
+| zone | n/a | `string` | `"us-central1-a"` | no |
+| machine_type | n/a | `string` | `"f1-micro"` | no |
+| database_instance_tier | n/a | `string` | `"db-f1-micro"` | no |
+| packagecloud_token | n/a | `string` | n/a | yes |
+| crypto_key | The crypto key to decrypt blobs | `string` | n/a | yes |
+| target_bucket | The name of the bucket where blobs should be migrated to | `string` | n/a | yes |
+| sql_dump_bucket | The name of the bucket where sql dump of tf_blobs table is stored | `string` | n/a | yes |
+| sql_dump_file_name | The name of sql dump file | `string` | n/a | yes |
+| max_allowed_packet | n/a | `int` | `131072` | no |
```
Comment (lbn) [inline modules/google/README.md]: Need to clarify that it's in bytes.

**Review by lbn — CHANGES_REQUESTED** [review]
Newlines also missing from a few files.

File: modules/google/main.tf [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+  deletion_protection = false
+
+  depends_on = [
+    google_compute_instance.node
+  ]
+}
+
+resource "google_sql_database" "database" {
+  name = "default"
+  instance = google_sql_database_instance.mysql.name
```
Comment (emocharnik) [inline modules/google/main.tf:71]: depends on

File: modules/google/main.tf [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+}
+
+locals {
+  ssh_user = "gce-blobtruck-user"
+  ssh_port = 22
+  ssh_timeout = var.ssh_timeout
+  ssh_private_key = tls_private_key.ssh.private_key_pem
+  ssh_public_key = tls_private_key.ssh.public_key_openssh
+
+  db_user = "blobtruck"
```
Comment (emocharnik) [inline modules/google/main.tf:24]: why this variable is hardcoded?

File: modules/google/main.tf (hunk 4 above)
Comment (brezitskiy) [reply modules/google/main.tf]: @lbn looks like it couldn't, there are circular dependencies
`data.template_file.blobtruck_yml` refers to `google_sql_database_instance.mysql` and `google_sql_database_instance.mysql` itself refers to `google_compute_instance.node`

**Review by lbn — APPROVED** [review]

**Review by emocharnik — APPROVED** [review]

**Pull request merged**